File: InJava/app/src/main/java/com/barryzeha/quizquestion/view/QuizActivity.java
package com.barryzeha.quizquestion.view;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;

import android.content.Intent;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.Handler;
import android.os.Looper;
import android.widget.Button;

import com.barryzeha.quizquestion.R;
import com.barryzeha.quizquestion.databinding.ActivityQuizBinding;
import com.barryzeha.quizquestion.model.QuestionModel;
import com.barryzeha.quizquestion.model.common.Constants;
import com.barryzeha.quizquestion.model.common.Util;
import com.barryzeha.quizquestion.viewModel.QuestionViewModel;

import java.util.HashSet;
import java.util.Random;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class QuizActivity extends AppCompatActivity {
    private ActivityQuizBinding bind;
    private Button[] arrayButtons =null;
    private CountDownTimer timer;
    private QuestionViewModel questionViewModel;
    private HashSet<Integer> randomSet;
    private int correctAnswer=0, numOfQuestion=1, totalPoints=0;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bind=ActivityQuizBinding.inflate(getLayoutInflater());
        Util.loadUrl(R.drawable.question_background,bind.ivBackgroundQuestion,this);
        setContentView(bind.getRoot());
        randomSet=generateRandomArray();


        arrayButtons= new Button[]{bind.btnAnswer1, bind.btnAnswer2,bind.btnAnswer3};
        setUpViewModel();
        initTimer();
    }

    private void setUpViewModel() {
        questionViewModel= new ViewModelProvider(this).get(QuestionViewModel.class);
        questionViewModel.getQuestionRandom(randomSet).observe(this, this::populateQuestionView);
    }
    private void populateQuestionView(QuestionModel question){
        bind.tvQuestion.setText(question.getQuestion());
        Util.loadUrl(question.getUrl(), bind.ivQuestion, this);
        bind.btnAnswer1.setText(question.getAnswer1());
        bind.btnAnswer2.setText(question.getAnswer2());
        bind.btnAnswer3.setText(question.getAnswer3());
        correctAnswer=question.getCorrectAnswer();
    }
    private void initTimer() {
        timer = new CountDownTimer(31000,1000) {
            @Override
            public void onTick(long millis) {
                long seconds = millis /1000;
                bind.pbTimer.setProgress(((int)seconds * 100) / 30);
                switch((int) seconds){
                    case 20:
                        bind.pbTimer.setIndicatorColor(ContextCompat.getColor(QuizActivity.this,R.color.green));
                        break;
                    case 15:
                        bind.pbTimer.setIndicatorColor(ContextCompat.getColor(QuizActivity.this,R.color.yellow));
                        break;
                    case 8:
                        bind.pbTimer.setIndicatorColor(ContextCompat.getColor(QuizActivity.this,R.color.red));
                        break;
                    default:
                        break;
                }
                bind.tvTimer.setText(String.valueOf(seconds));
            }

            @Override
            public void onFinish() {
                timer.cancel();
                arrayButtons[correctAnswer -1].setBackgroundColor(ContextCompat.getColor(QuizActivity.this,R.color.green));
                enableButtons(false);
            }
        };
        setUpButtonListeners();
        timer.start();
    }

    private void setUpButtonListeners() {
        for(Button btn : arrayButtons){
            btn.setOnClickListener(v->{
                timer.cancel();
                if(Integer.parseInt(btn.getTag().toString())==correctAnswer){
                    //le cambiamos el color del botón si es correcta la respuesta
                    btn.setBackgroundColor(ContextCompat.getColor(QuizActivity.this,R.color.green));
                    totalPoints += Integer.parseInt(bind.tvTimer.getText().toString());
                }
                else{
                    //color rojo al botón si la respuesta es errónea
                    btn.setBackgroundColor(ContextCompat.getColor(QuizActivity.this,R.color.red));
                    //marcamos el botón con la respuesta correcta con el color verde
                    arrayButtons[correctAnswer -1].setBackgroundColor(ContextCompat.getColor(QuizActivity.this,R.color.green));

                }
                enableButtons(false);
                nextQuestion();
            })
            ;
        }
    }
    private void nextQuestion(){
        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            if(numOfQuestion==10){
               startActivity(new Intent(QuizActivity.this, ScoreActivity.class).putExtra(Constants.POINTS_EXTRAS,totalPoints));
            }
            else{
                questionViewModel.getQuestionRandom(randomSet);
                //le volvemos a poner el color inicial a los botones
                for(Button btn :arrayButtons){
                    btn.setBackgroundColor(ContextCompat.getColor(QuizActivity.this,android.R.color.transparent));
                }
                //reiniciamos el color del progress Indicator  del timer
                bind.pbTimer.setIndicatorColor(ContextCompat.getColor(this,R.color.white));
                enableButtons(true);
                initTimer();
                numOfQuestion++;
            }
        },1000);
    }
    private void enableButtons(boolean state) {
        bind.btnAnswer1.setEnabled(state);
        bind.btnAnswer2.setEnabled(state);
        bind.btnAnswer3.setEnabled(state);
    }
    private HashSet<Integer> generateRandomArray(){
        int size=10;
        Random random= new Random();
        HashSet<Integer> s= new HashSet<>(size);
        while(s.size() < size){
            s.add(random.nextInt(27));
        }
        return s;
    }
}

File: InKotlin/app/src/main/java/com/barryzeha/quizquestion/view/QuizActivity.kt
package com.barryzeha.quizquestion.view

import android.os.Bundle
import android.os.CountDownTimer
import android.os.Handler
import android.os.Looper
import android.widget.Button
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import com.barryzeha.quizquestion.R
import com.barryzeha.quizquestion.common.*
import com.barryzeha.quizquestion.databinding.ActivityQuizBinding
import com.barryzeha.quizquestion.model.QuestionModel
import com.barryzeha.quizquestion.viewModel.QuestionViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlin.random.Random

@AndroidEntryPoint
class QuizActivity : AppCompatActivity() {
    private lateinit var bind: ActivityQuizBinding
    private lateinit var timer:CountDownTimer
    private lateinit var arrayButtons:Array<Button>
    private val viewModelQuestion:QuestionViewModel by viewModels()
    private lateinit var randomSet:HashSet<Int>
    private var answer=0
    private var numOfQuestion=1
    private var totalPoints=0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bind = ActivityQuizBinding.inflate(layoutInflater)
        setContentView(bind.root)
        randomSet=generateRandomArray()
        arrayButtons = arrayOf(bind.btnAnswer1,bind.btnAnswer2,bind.btnAnswer3)
        bind.ivBackgroundQuestion.loadUrl(R.drawable.question_background)
        setUpViewModel()
        initTimer()


    }
    private fun setUpViewModel(){
        //first call

        viewModelQuestion.getRandomQuestion(randomSet).observe(this, Observer(::populateQuestionView))
    }
    private fun populateQuestionView(question:QuestionModel?){
        question?.let{q->
            bind.tvQuestion.text=q.question
            bind.ivQuestion.loadUrl(q.url)
            bind.btnAnswer1.text=q.answer1
            bind.btnAnswer2.text=q.answer2
            bind.btnAnswer3.text=q.answer3
            answer=q.correctAnswer
        }
    }

    private fun initTimer(){
        timer = object : CountDownTimer(31000,1000){
            override fun onTick(millis: Long) {

                val seconds= millis / 1000
                bind.pbTimer.progress=(seconds.toInt() * 100) / 30
                when(seconds){
                    20L->bind.pbTimer.setIndicatorColor(ContextCompat.getColor(this@QuizActivity,R.color.green))
                    15L-> bind.pbTimer.setIndicatorColor(ContextCompat.getColor(this@QuizActivity,R.color.yellow))
                    8L-> bind.pbTimer.setIndicatorColor(ContextCompat.getColor(this@QuizActivity,R.color.red))
                }
                bind.tvTimer.text=seconds.toString()
            }

            override fun onFinish() {
                timer.cancel()
                arrayButtons[answer - 1].setBackgroundColor(ContextCompat.getColor(this@QuizActivity,R.color.green))
                enableButtons(false)
                nextQuestion()
            }

        }
        setUpButtonsListeners()
        timer.start()
    }
    private fun setUpButtonsListeners(){
        arrayButtons.forEach { btn->
            btn.setOnClickListener {
                timer.cancel()
                if(btn.tag.toString().toInt()== answer){
                    btn.setBackgroundColor(ContextCompat.getColor(this,R.color.green))
                    totalPoints +=bind.tvTimer.text.toString().toInt()
                }
                else{
                    btn.incorrectAnswer(arrayButtons,answer)
                }
                enableButtons(false)
                nextQuestion()
            }
        }
    }
    private fun nextQuestion(){
        Handler(Looper.getMainLooper()).postDelayed(
            {
                if(numOfQuestion ==10){

                    startActivity<ScoreActivity> {
                       putExtra(Constant.POINTS_EXTRAS,totalPoints)
                    }
                }
                else{
                    viewModelQuestion.getRandomQuestion(randomSet)
                    arrayButtons.forEach { it.clear() }
                    bind.pbTimer.setIndicatorColor(ContextCompat.getColor(this,R.color.white))
                    enableButtons(true)
                    initTimer()
                    numOfQuestion++

                }
            },1000
        )
    }
    private fun enableButtons(state:Boolean){
        bind.btnAnswer1.isEnabled=state
        bind.btnAnswer2.isEnabled=state
        bind.btnAnswer3.isEnabled=state
    }
    private fun generateRandomArray():HashSet<Int>{
        val size=10
        val s= HashSet<Int>(size)
        while(s.size < size){
            s += Random.nextInt(0,27)
        }
        return s
    }
}